domain/application/state/application.go

```diff
@@ -132,6 +134,21 @@ func (st *State) CreateApplication(ctx domain.AtomicContext, name string, app ap
 		}
 	}
 
+	// build resources to add.
+	resources, err := st.buildResourceToAdd(
+		appDetails.ApplicationID.String(),
+		appDetails.CharmID,
+		app.Resources,
+		app.Charm.Metadata.Resources)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	// Prepare insert resources statement.
+	insertApplicationResources, err := st.prepareResourceInsert(resources)
```

Comment:
With the `StateBase` caching prepared statements, you could do this preperation inside a function called from within the transaction. The cost isn't so high and it might be a bit neater. It also seems to be the pattern used elsewhere in this file.

Reply:
Oh, I wasn't aware of that.